### PR TITLE
Implementing the ``chmod`` support

### PR DESCRIPTION
**Pull request by webbunivAdmin**
### Add `chmod` Support to Eio

This PR implements support for changing file permissions using the `chmod` operation in the Eio library.
#764 

#### Key Changes:
- **Added `chmod` function** that allows setting file permissions on a target resource:
  - **Parameters:**
    - `~follow`: Determines whether to follow symlinks.
    - `~perm`: Specifies the permission bits to set.
    - `t`: The target file or directory.
- **Exception Handling:**
  - Exceptions are handled to provide robust error reporting.
  - Context-aware re-raising of exceptions in case of failure for better debugging.
- **Seamless Integration** with Eio’s resource abstraction, supporting various permission modes.

#### Tests:
-

### CODE REVIEW

**Review by patricoferris — CHANGES_REQUESTED** [review]
Thanks @webbunivAdmin !

For future reference, it would be good to correctly attribute code to the original author. I believe most of this was from the branch I pointed to. You could have [committed using co-authored-by](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) or by [cherry-picking](https://www.atlassian.com/git/tutorials/cherry-pick) from my branch.

I think we're going to need some tests here too, similar to those in the symlink PR

File: examples/chmod/main.ml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  Switch.run @@ fun sw ->
+    let fs = Eio.Stdenv.fs env in
+    let tmp_dir = Eio.Path.(fs / "tmp") in
+    let file_path = Eio.Path.(tmp_dir / "test_file") in
+    Eio.Path.(open_out ~sw ~create:(`If_missing 0o600) file_path) |> Eio.Flow.close;
+
+    let permissions = get_permissions (Eio.Path.native_exn file_path) env in
+    Printf.printf "Permissions: %o\n" permissions;
+
+    Eio.Path.chmod ~follow:true ~perm:0o600 file_path;
```
Comment (patricoferris) [inline examples/chmod/main.ml:19]: I think we could remove this example and make the code a test in `tests/fs.md`. Also it would be good to change the permissions to something other that the same permissions used to create the file, that way we can see `chmod` working.

**Review by patricoferris — CHANGES_REQUESTED** [review]
Also seems to be a `tmp/test_file` that shouldn't be committed.

File: tests/fs.md [hunk 2]
```diff
@@ -18,7 +18,6 @@ open Eio.Std
 let ( / ) = Path.( / )
 
 let run ?clear:(paths = []) fn =
-  Eio_main.run @@ fun env ->
```
Comment (patricoferris) [inline tests/fs.md:21]: Why did you remove this @webbunivAdmin ? It is very important as it is the effect handler for Eio code. Without this none of this tests will work.

File: tests/fs.md [hunk 3]
```diff
@@ -1011,3 +1017,23 @@ Exception: Failure "Simulated error".
 +"/" / "" = "/"
 - : unit = ()
 ```
+
+```ocaml
```
Comment (patricoferris) [inline tests/fs.md:1021]: To run the test make sure you run `dune runtest` and `dune promote` the output back to this file.

**Review by patricoferris — CHANGES_REQUESTED** [review]
Thanks @webbunivAdmin -- getting there :))

**Comment by webbunivAdmin** [comment]
@patricoferris for checkout for the new tests have written

**Comment by webbunivAdmin** [comment]
@patricoferris for checkout for the new tests have written

**Comment by patricoferris** [comment]
Hey @webbunivAdmin,

A few notes from me here. Firstly, in terms of co-authoring, you only need to add this to your commits that are indeed co-authored by me. Everything after the initial commit is your own work and need not have co-authored on them.

Secondly, I do wonder what these pushes of `update fs.md` are for? Do you not have the code building and running locally to be able to run `dune runtest` to see if it is working _before_ pushing. It is also possible to undo commits on _your_ branch:

```sh
git reset --hard HEAD~n        # Reset head backwards n commits, dropping the commits
edit edit edit
git add foo                    # add new changes
git commit -m "Useful message" # commit them
git push -f                    # force push to your branch to undo the commits
```

Some of the failures right now are not due to the code (just flakey tests/CI). However, the latest code still has no test of the `chmod` feature.

File: tests/fs.md [hunk 4]
```diff
@@ -1011,3 +1020,32 @@ Exception: Failure "Simulated error".
...
+
+  let file_path = cwd / "test-file" in
+  Path.save ~create:(`Exclusive 0o644) file_path "test data";
+  traceln "+create <cwd:test-file> with permissions 0o644 -> ok";
+
+  let initial_perm = (Path.stat ~follow:true file_path).perm in
+  traceln "+<cwd:test-file> initial permissions = %o" initial_perm;
+  assert (initial_perm = 0o644);
+
+  Path.chmod ~follow:true ~perm:0o400 file_path;
```
Comment (patricoferris) [inline tests/fs.md:1037]: I envisage this code to look something like:

```ocaml
try_chmod ~follow:true ~perm:0o400 file_path;
try_stat ~info_type:`Perm file_path;
```

Note:

1. It uses `try_chmod` which prints the information about the execution of `chmod`.
2. It uses `try_stat` to print information about the file, I've modified it slightly so that `try_stat` would look something like 
```diff
-let try_stat path =
+let try_stat ?(info_type=`Kind) path =
   let stat ~follow =
-    match Eio.Path.stat ~follow path with
-    | info -> Fmt.str "@[<h>%a@]" Eio.File.Stat.pp_kind info.kind
+    match Eio.Path.stat ~follow path, info_type with
+    | info, `Perm -> Fmt.str "@[<h>%o@]" info.perm
+    | info, `Kind -> Fmt.str "@[<h>%a@]" Eio.File.Stat.pp_kind info.kind
     | exception Eio.Io (e, _) -> Fmt.str "@[<h>%a@]" Eio.Exn.pp_err e
   in
```

File: lib_eio_windows/fs.ml [hunk 5]
```diff
@@ -176,6 +181,7 @@ end = struct
     with_parent_dir t path @@ fun dirfd path ->
     Err.run (Low_level.symlink ~link_to dirfd) path
 
+
```
Comment (patricoferris) [inline lib_eio_windows/fs.ml:184]: Make sure your commits don't have unnecessary changes like this newline.

File: lib_eio_windows/fs.ml [hunk 6]
```diff
@@ -89,6 +89,11 @@ end = struct
 
   let v ~label ~sandbox dir_path = { dir_path; sandbox; label; closed = false }
 
+  let chmod (_t : t) ~(follow : bool) ~(perm : int) (_path : string) : unit =
```
Comment (patricoferris) [inline lib_eio_windows/fs.ml:92]: Like with the other functions that do not have an implementation, we should actually still create a stub in `eio_windows_stubs.c` that fails. See, for example, how symlink does it. This reduces the burden for someone to come along and do a `chmod` implementation for Windows.

**Comment by webbunivAdmin** [comment]
@patricoferris At last i made

**Comment by patricoferris** [comment]
Before reviewing @webbunivAdmin could you tidy up the Git history a little here. 35 of the 49 commits are called `Update fs.md` ^^" ! Mostly you could just squash the history with a soft reset and a force push to your branch.

**Comment by webbunivAdmin** [comment]
@patricoferris i thing am done

**Comment by patricoferris** [comment]
Hey @webbunivAdmin, sorry if I wasn't clear enough in my previous post. To make reviewing this easier could you tidy up the git history please. For example there is a pretty good stackoverflow post about how you can do that here: https://stackoverflow.com/questions/5189560/how-do-i-squash-my-last-n-commits-together 

Does that make sense? If you don't want to squash the history do let me know and I'll take over the PR at some point to get it in a reviewable shape. Thanks :))

**Comment by patricoferris** [comment]
Closing this in favour of #785
